engine/admin/multi-service_container.md

```diff
@@ -0,0 +1,91 @@
+---
+description: How to run more than one process in a container
+keywords: docker, supervisor, process management
+redirect_from:
+- /engine/articles/using_supervisord/
+- /engine/admin/using_supervisord/
+title: Run multiple services in a container
+---
+
+A container's main running process is the `ENTRYPOINT` and/or `CMD` at the
+end of the `Dockerfile`. It is generally recommended that you separate areas of
+concern by using one service per container. That service may fork into multiple
+processes (for example, Apache web server starts multiple worker processes).
+It's ok to have multiple processes, but to get the most benefit out of Docker,
+avoid one container being responsible for multiple aspects of your overall
+application. You can connect multiple containers using user-defined networks and
+shared volumes.
+
+The container's main process is responsible for managing all processes that it
+starts. In some cases, the main process isn't well-designed, and doesn't handle
+"reaping" (stopping) child processes gracefully when the container exists. If
+your process falls into this category, you can use the `--init` option when you
+run the container. The `--init` flag inserts a tiny init-process into the
+container as the main process, and handles reaping of all processes when the
+container exits. Handling such processes this way is superior to using a
+full-fledged init process such as `sysvinit`, `upstart`, or `systemd` to handle
+process lifecycle within your container.
+
+If you need to run more than one service within a container, you can accomplish
+this in a few different ways.
+
+- Put all of your commands in a wrapper script, complete with testing and
+  debugging information. Run the wrapper script as your `CMD`. This is a very
+  naive example. First, the wrapper script:
+
+  ```bash
+  #!/bin/bash
+
+  # Start the first process
+  ./my_first_process -D
+  status=$?
+  if [ $status -ne 0 ]; then
+    echo "Failed to start my_first_process: $status"
+    exit $status
+  fi
+
+  # Start the second process
+  ./my_second_process -D
+  status=$?
+  if [ $status -ne 0 ]; then
+    echo "Failed to start my_second_process: $status"
+    exit $status
+  fi
+
+  # Naive check to see if either of them exited immediately
+  PROCESS_!_STATUS=$(ps aux |grep -q my_first_process)
+  PROCESS_2_STATUS=$(ps aux |grep -q my_second_process)
+  if [ $PROCESS_!_STATUS || $PROCESS_2_STATUS ];
+    echo "One of the processes has already exited."
+    exit -1
+  fi
+  ```
```

Comment:
This script, as written, will exit immediately and thus PID 1 will exit and the two services started will be killed.  Is that the intention for this example?  The hacky way I usually see folks solve this is by putting this `ps` logic into a loop or using something silly like `tail -f` to just keep the container alive (`tail -f /dev/null` is the cheekiest example of "keep this container from dying" that I've seen to date :smile:).